src/types/mod.rs

```diff
@@ -56,11 +56,30 @@ macro_rules! pyobject_native_type_named (
     };
 );
 
+macro_rules! impl_layout {
+    ($name: ty, $layout: path) => {
+        unsafe impl $crate::type_object::PyObjectLayout<$name> for $layout {
+            unsafe fn unchecked_ref(&self) -> &$name {
+                &*((&self) as *const &Self as *const _)
+            }
+            unsafe fn unchecked_mut(&self) -> &mut $name {
+                &mut *((&self) as *const &Self as *const _ as *mut _)
+            }
+        }
+    };
+}
+
 #[macro_export]
 macro_rules! pyobject_native_type {
     ($name: ty, $layout: path, $typeobject: expr, $module: expr, $checkfunction: path $(,$type_param: ident)*) => {
-        impl $crate::type_object::PyObjectLayout<$name> for $layout {}
+        impl_layout!($name, $layout);
         impl $crate::type_object::PyObjectSizedLayout<$name> for $layout {}
+        impl $crate::derive_utils::PyBaseTypeUtils for $name {
+            type Dict = $crate::pyclass_slots::PyClassDummySlot;
+            type WeakRef = $crate::pyclass_slots::PyClassDummySlot;
+            type Layout = $crate::pycell::PyCellBase<$name>;
```

Comment:
Do native types need borrow checking? I would have thought that they should just continue to be accessed with `&PyAny`.

Reply:
It's confusing but `<T as PyBaseTypeUtils>::Layout` means *the layout as a baseclass of PyClass`.
Maybe we need a clearer name.